src/compiler/crystal/semantic/cleanup_transformer.cr

```diff
@@ -510,6 +522,14 @@ module Crystal
         arg.expressions.each do |exp|
           check_arg_is_not_closure(node, message, exp)
         end
+      when Call
+        # detect closured vars in a call to Proc.new(&block)
+        if arg.name == "new" && (block = arg.block) && (obj_type = arg.obj.try &.type?)
```

Comment:
I was thinking we could check here whether this is the std definition and do this extra logic, otherwise don't do it, but still allow refefinitions of Proc.new. But I guess this is fine. That said, someone could add an overload to new and this check might break. So we better also check that we are only doing this for the std definition.

Reply:
One difficulty is that it's currently not possible to tell which `Proc.new` def is "the" stdlib definition; it is simply the first definition the compiler encounters. So the safer way is to simply check the body of _every_ definition of `Proc.new(&)`, as done elsewhere in this file, even if that means no other overloads of `Proc.new` can take a block. This technique is inapplicable to other magic methods like `String.interpolation` though.

IMO the stdlib needs a way to signal to the users that some defs are involved in compiler magic and therefore should never be redefined or overloaded unless they "know what they're doing"; redefining them is undefined behaviour. Then the compiler can pretend that said methods are precisely the stdlib ones.

Reply:
Oh, with std definition I meant the check you are doing now in the other place.

Reply:
I mean, if someone redefined new with a different body then everything will still work, except that the compile time check for this will be done at runtime.

What I don't want is for someone to overload Proc.new with a different body, and this code working on that because it might be incorrect. So this check must only be done if the body is just the block arg.

Reply:
If we proceed that way, probably _any_ call that has the same body (just the captured block arg and nothing else) should trigger the compile-time check too, not just `Proc.new(&)`.

Reply:
Sounds good.